### PR TITLE
tests: unit tests for workflows/tasks/submission

### PR DESCRIPTION
**Pull request by rikirenz**
Addresses #1773

- [x] `submit_rt_ticket`
- [x] `create_ticket`
- [x] `reply_ticket`
- [x] `close_ticket`
- [x] `send_robotupload`
- [x] `add_note_entry`
- [x] `filter_keywords`
- [x] `prepare_keywords`
- [x] `user_pdf_get`
- [x] `prepare_files`
- [x] `remove_references`

### CODE REVIEW

File: tests/integration/workflows/test_submission_workflow.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    small_app.config['PRODUCTION_MODE'] = True
+
+    assert submit_rt_ticket(obj,
+                            queue,
+                            context.get('subject'),
+                            body,
+                            'mail@ma.il',
+                            ticket_id_key)
+
+    assert mocker_rt_instance.call_count == 1
```
Comment (jacquerie) [inline tests/integration/workflows/test_submission_workflow.py]: Take a look at this: https://testing.googleblog.com/2015/01/testing-on-toilet-change-detector-tests.html

Comment (rikirenz) [inline tests/integration/workflows/test_submission_workflow.py]: if I remove the line: `assert mocker_rt_instance.call_count == 1`. is it good for you ? or should I do other changes?

**Comment by rikirenz** [comment]
@jacquerie I just realized that I put the test in the wrong folder. This is a unit test not a integration test. I removed the line that was checking that the `get_instance()` method was triggered. BTW I did not understand your comment (about the google tests). Could you give me more explanations?

File: tests/unit/workflows/test_workflows_tasks_submission.py [hunk 2]
```diff
@@ -35,9 +60,40 @@ class DummyEng(object):
...
+            side_effect=mock_get_rt_instance)
+def test_submit_rt_ticket(mocker_rt_instance):
+    """test that a ticket is sent to the RT service"""
+
+    body = ''
+    queue = ''
+    context = {}
+    obj = MockObj(1, {}, {})
+    ticket_id_key = 'ticket_id'
+    current_app.config['PRODUCTION_MODE'] = True
```
Comment (rikirenz) [inline tests/unit/workflows/test_workflows_tasks_submission.py]: I have added this line to simulate the production mode. I have tried to use the `app` fixture in the `conftest.py` file but I had problems. Is it ok in this way?

Comment (jacquerie) [inline tests/unit/workflows/test_workflows_tasks_submission.py]: I believe not. This is modifying the configuration for all the tests that follow, potentially introducing a dependency between tests.

Comment (jacquerie) [inline tests/unit/workflows/test_workflows_tasks_submission.py]: This is the correct way to do that: https://github.com/inspirehep/inspire-next/blob/cef4089e2c381f00bcd0631e9f388db30b6bb680/tests/integration/workflows/test_arxiv_workflow.py#L271

Comment (jacquerie) [inline tests/unit/workflows/test_workflows_tasks_submission.py]: And this is another way to do it, but I like it less, despite the fact that I just wrote it ._. https://github.com/inspirehep/inspire-next/blob/cef4089e2c381f00bcd0631e9f388db30b6bb680/tests/unit/workflows/test_workflows_tasks_actions.py#L164-L176

File: tests/unit/workflows/test_workflows_tasks_submission.py [hunk 3]
```diff
@@ -35,9 +60,32 @@ class DummyEng(object):
...
+
+
+@mock.patch('inspirehep.modules.workflows.tasks.submission.get_instance', side_effect=mock_get_rt_instance)
+@mock.patch('inspirehep.modules.workflows.tasks.submission.in_production_mode', side_effect=mock_in_production_mode)
+def test_submit_rt_ticket(mock_production_mode, mocker_rt_instance):
+    obj = MockObj(1, {}, {})
+    current_app.config['PRODUCTION_MODE'] = True
+
+    assert submit_rt_ticket(obj, '', None, '', 'mail@ma.il', 'ticket_id')
+    assert obj.extra_data['ticket_id'] == 1
```
Comment (rikirenz) [inline tests/unit/workflows/test_workflows_tasks_submission.py]: I have not found a better way to test this function. I think is useless.

File: tests/unit/workflows/test_workflows_tasks_submission.py [hunk 4]
```diff
@@ -22,22 +22,55 @@
...
+            self.data = data
+            self.extra_data = extra_data
+
+    mock_file = MagicMock()
+    mock_file.create_ticket = MagicMock(return_value=1)
+    mock_rt_instance.return_value = mock_file
+    mock_production_mode.return_value = True
+
+    obj = MockObj(1, {}, {})
+    current_app.config['PRODUCTION_MODE'] = True
```
Comment (rikirenz) [inline tests/unit/workflows/test_workflows_tasks_submission.py]: this line is not necessary

File: tests/unit/workflows/test_workflows_tasks_submission.py [hunk 5]
```diff
@@ -22,22 +22,380 @@
...
+            self._log = StringIO()
+
+        def info(self, *args, **kwargs):
+            self._log.write(args[0])
+
+        def error(self, *args, **kwargs):
+            self.info(args[0])
+
+        def debug(self, message, format=None):
+            self.info(message)
```
Comment (jacquerie) [inline tests/unit/workflows/test_workflows_tasks_submission.py]: Uhm, these three logging methods have all different signatures and implementations, despite being (essentially) the same thing....

Also, with this design one cannot see if the proper logging level happened. Maybe we should have three `StringIO` objects for the three different log levels? Maybe we should add some kind of marker (like `DEBUG:`) in front?

File: tests/unit/workflows/test_workflows_tasks_submission.py [hunk 6]
```diff
@@ -22,22 +22,380 @@
...
+def mock_context_function(user, obj):
+    return {}
+
+
+@mock.patch('inspirehep.modules.workflows.tasks.submission.get_instance')
+@mock.patch('inspirehep.modules.workflows.tasks.submission.in_production_mode')
+def test_submit_rt_ticket(mock_production_mode, mock_rt_instance):
+    mock_file = MagicMock()
+    mock_file.create_ticket = MagicMock(return_value=1)
+    mock_rt_instance.return_value = mock_file
```
Comment (jacquerie) [inline tests/unit/workflows/test_workflows_tasks_submission.py]: I would expect that a variable called `mock_file` is the mock of a file, but you clearly mean something different. What do you mean here?

File: tests/unit/workflows/test_workflows_tasks_submission.py [hunk 7]
```diff
@@ -22,22 +22,380 @@
...
+    mock_production_mode.return_value = True
```
Comment (jacquerie) [inline tests/unit/workflows/test_workflows_tasks_submission.py]: Can be pulled up in the mock declaration, or avoided entirely (Classicist vs. Mockist).

File: tests/unit/workflows/test_workflows_tasks_submission.py [hunk 8]
```diff
@@ -22,22 +22,380 @@
...
+
+
+@pytest.mark.parametrize('extra_data_key', [None, 'something'])
+@mock.patch('inspirehep.modules.workflows.tasks.submission.in_production_mode')
+@mock.patch('inspirehep.utils.robotupload.make_robotupload_marcxml')
+@mock.patch('inspirehep.dojson.utils.legacy_export_as_marc')
+def test_send_robotupload(mock_legacy_export_as_marc,
+                          mock_make_robotupload_marcxml,
+                          mock_production_mode,
+                          extra_data_key):
```
Comment (jacquerie) [inline tests/unit/workflows/test_workflows_tasks_submission.py]: Tip: when mock names become this unwieldy you can just use the initials of every letter: `mock_legacy_export_as_marc` -> `l_e_a_m`.

Although take note that, when that happens, maybe you're mocking too many things...

File: tests/unit/workflows/test_workflows_tasks_submission.py (hunk 6 above)
Comment (rikirenz) [inline tests/unit/workflows/test_workflows_tasks_submission.py]: It is wrong. The correct name should be `mock_ticket_obj`

File: tests/unit/workflows/test_workflows_tasks_submission.py (hunk 7 above)
Comment (rikirenz) [inline tests/unit/workflows/test_workflows_tasks_submission.py]: Do you mean: create a specific app object that simulates the production mode?

File: tests/unit/workflows/test_workflows_tasks_submission.py (hunk 5 above)
Comment (rikirenz) [inline tests/unit/workflows/test_workflows_tasks_submission.py]: :+1:

File: tests/unit/workflows/test_workflows_tasks_submission.py (hunk 8 above)
Comment (rikirenz) [inline tests/unit/workflows/test_workflows_tasks_submission.py]: :+1:

**Comment by jacquerie** [comment]
Have a look at https://github.com/inspirehep/inspire-next/blob/b909f11eac21efe5b3bbe9f667337d57bb6a7423/tests/unit/workflows/test_workflows_tasks_submission.py to see how you can improve these tests.

**Comment by rikirenz** [comment]
@jacquerie  Anyway after several months of tests I understood your point. When We will have time I will try to do tests in the classicist way.

File: tests/unit/workflows/test_workflows_tasks_submission.py (hunk 8 above)
Comment (david-caro) [inline tests/unit/workflows/test_workflows_tasks_submission.py]: I do prefer explicit names than the shortened ones, even if they are 'unwieldy', well, specially if they are 'unwieldy' as it's then when you are most probably going to be confused.

**Comment by jacquerie** [comment]
Closed in favor of #2313

**Pull request closed without merging**